redhat-appstudio/application-service
# [DEVHAS-100] Add watch support for Environment CR with the Binding controller

### Description:
This PR adds the watch support for the Environment CR in the Binding controller. When an Environment CR is updated, the Binding controller regenerates the GitOps resources and commits them if any change is present. 

When a Binding CR is created or reconciled, we also check for the Environment CR specified in the Binding spec. For application-service controller, only the Environment CR spec configuration for ENV is of interest ie; `spec.configuration.env`. If `spec.configuration.env` is present, we check if an env of the same name exist in the binding component spec ie; `spec.components[*].configuration.env`. If one doesn't exist, the `spec.configuration.env` is appended for the component GitOps deployment resource env, skipped if it already exists in `spec.components[*].configuration.env`.

### To test it out:
- create app `k create -f application/app.yaml`
- create component(s) component-sample and component-sample2 as per binding spec `k create -f component/component-basic.yaml`
- create snapshot `k apply -f snapshot/snapshot.yaml`
- create environment `k apply -f environment/environment.yaml`
- create binding `k apply -f binding/binding.yaml`
- check and verify the env in deployment-patch.yaml for both the components
- update the env value in `environment/environment.yaml` and apply again `k apply -f environment/environment.yaml`
- all the bindings referencing the environment will get reconciled, verify the env update in the GitOps repo again

Happy reviewing! 😄
